kubetest2/pkg/app/cmd.go

```diff
@@ -146,6 +167,58 @@ type options struct {
 	artifacts string
 }
 
+func newOptions(name, deployerUsage string) *options {
+	opt := &options{
+		name:          name,
+		deployerUsage: deployerUsage,
+	}
+
+	flags := pflag.NewFlagSet(name, pflag.ContinueOnError)
+	// unknown flags are forwarded to the deployer as arguments
+	flags.ParseErrorsWhitelist.UnknownFlags = true
+
+	// register all first class kubetest2 flags
+	flags.BoolVarP(&opt.help, "help", "h", false, "display help")
+	flags.BoolVar(&opt.build, "build", false, "build kubernetes")
+	flags.BoolVar(&opt.up, "up", false, "provision the test cluster")
+	flags.BoolVar(&opt.down, "down", false, "tear down the test cluster")
+	flags.StringVar(&opt.test, "test", "", "test type to run, if unset no tests will run")
```

Comment:
so, kubetest2 will be able to call a deployer's `--test`?
should all deployers re-define the `--up, --down, --build` etc flags like so?

Reply:
deployer _binaries_ define these flags here, but the logic is inside this kubetest2app package that they import, they don't need to redefine it.

flags / args that are not defined here are forwarded to the deployer as arguments for processing.

anything after the bare `--` is passed to the tester, selected by the `--test` flag.

the process tree looks like
```
`kubetest2 kind --test=exec -- echo "foo"`
(passes everything but "kind" to the `kubetest2-kind` deployer specific binary, including env, signals)
-> 
  `kubetest2-kind --text=exec --echo "foo"`
  (parses test = "exec", passes []string{} to deployer, and `[]string{"echo" "\"foo\""}` to testers["exec"]
    -> 
     (exec tester invokes `TesterArgs[0] TesterArgs[1:]...` and passes signals etc)
        `echo "foo"`
````

Reply:
kubetest2 will only have a handful of 100% common flags here, like `help`, and the flags that specify the high level flow (build, up, down, test)

things like build options are deployer dependent.

tester flags are after the `--` and depend on `--test=footester`, that way EG ginkgo flags are only for ginkgo testing.

Reply:
oh, my mistake i tough this was under the kind deployer, but it was under pkg/app. :(

Reply:
gotcha. 

FWIW I also intend to write fairly thorough docs for this, specifically including expectations for deployers and testers once we have at least one working deployer.

This is definitely a little confusing either way... but ideally deployer authors etc. will never need to look at all this command line boilerplate, they'll just need to look at one of the other deployers.

Reply:
no, i don't think it's confusing.
but yes, existing deployers will be a stepping stone for writing new ones, for sure.